### PR TITLE
Fixes #16595 - Add directives to scoped migrations

### PR DESCRIPTION
**Pull request by ShimShtein**
PR to implement https://github.com/theforeman/rfcs/pull/15

### CODE REVIEW

File: db/migrate/20990101999999_sample_migration.foreman_plugin_template.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+class SampleMigration < ActiveRecord::Migration
```
Comment (domcleal) [inline db/migrate/20990101999999_sample_migration.foreman_plugin_template.rb]: Don't add migrations to this template, this does nothing, and will not be removed by authors.

File: README.md [hunk 2]
```diff
@@ -47,6 +47,27 @@ development team below.
...
 
+## Migrations
+
+Your migrations should be placed under `db/migrate` folder. Every migration file
+should be suffixed by the name of the plugin, for example:
+```
+db/migrate/20990101999999_add_my_table.foreman_plugin_example.rb
+```
+
+### Migrating your plugin only:
```
Comment (domcleal) [inline README.md]: This documentation should be in the plugin docs, not the template docs.

File: db/migrate/20990101999999_sample_migration.foreman_plugin_template.rb (hunk 1 above)
Comment (ShimShtein) [inline db/migrate/20990101999999_sample_migration.foreman_plugin_template.rb]: Where can I give examples to this? Do we have to rely only on docs for it?

My long-term suggestion would be a generator, but I don't think it's the right time to do that.

**Comment by ShimShtein** [comment]
I have split the documentation: the explanation how to name migrations I have left in template docs (it's for devs only) and the directions about using the migrations were moved to plugin docs (the user will benefit most from them)

File: db/migrate/20990101999999_sample_migration.foreman_plugin_template.rb (hunk 1 above)
Comment (domcleal) [inline db/migrate/20990101999999_sample_migration.foreman_plugin_template.rb]: Yes, docs are for examples. This ought to be an empty, usable plugin - not every plugin should add a new, empty migration.

File: README.plugin.md [hunk 3]
```diff
@@ -7,6 +7,18 @@
 See [How_to_Install_a_Plugin](http://projects.theforeman.org/projects/foreman/wiki/How_to_Install_a_Plugin)
 for how to install Foreman plugins
 
+## Migrating your plugin:
```
Comment (domcleal) [inline README.plugin.md]: This is user documentation, it shouldn't say "your plugin", it isn't the user's plugin. This is also only applicable to source installations, it shouldn't be prominent in users docs.

Comment (ShimShtein) [inline README.plugin.md]: > it shouldn't be prominent in users docs.

That was my original thought. So what would you suggest? My last resort would be [wiki](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin).

Comment (domcleal) [inline README.plugin.md]: The wiki's the main source of detailed maintainer docs, it should probably be changed first. I'd suggest leaving it out of here entirely, I don't think it's important in the plugin README - the only bit that may be useful is uninstallation, but this probably doesn't implement the complete process?

**Comment by domcleal** [comment]
Also, the Redmine ticket is against Foreman - this repo doesn't have a Redmine project.

File: README.plugin.md (hunk 3 above)
Comment (ShimShtein) [inline README.plugin.md]: > The wiki's the main source of detailed maintainer docs.

Thanks, I'll put it there.

> this probably doesn't implement the complete process

Actually it's all that is needed - first remove the migrations as I mentioned and then you can remove the code dependency just as you have added it be it RPM, code or gem.
There are still pitfalls like "orphaned" settings e.t.c. but they should be handled by the plugins IMHO.
My only consideration is the way the user consumes the plugin: if the plugin is packaged - it should be possible to add a step that will perform the migration automatically, in any other case  (gem or code) the user is responsible for running the migration.

Comment (domcleal) [inline README.plugin.md]: So that documentation is available on the wiki currently, that's the best place to maintain it.

**Comment by lzap** [comment]
I think we should give an advice how to test uninstallations. We've seen many rollback failures as it's not well covered by tests. From now on, plugins should ship with a rake task that will test installation/uninstallation somehow. Either some kind of dummy migration and rollback or simply by migrating on an empty database.

**Comment by ShimShtein** [comment]
@lzap What about changing the jenkins job for plugins so, that it will perform the down migration, remove the plugin and run some smoke test that the system is still usable?

**Comment by lzap** [comment]
> @lzap What about changing the jenkins job for plugins so, that it will perform the down migration, remove the plugin and run some smoke test that the system is still usable? 

Sure, that's basically what I have in my mind, but I assume there must
be a rake task in order to do that for each plugin? Or we can create a
separate jenkins job that will do it for each plugin? Not sure.

-- 
Later,
 Lukas #lzap Zapletal

**Comment by ShimShtein** [comment]
I am not an expert in our jenkins jobs, but I can't see how can we do it in a rake task - we need to remove code dependency - for dev setups it would be removing the plugin from `gemfile.local.rb`, I am not sure what is the correct step for test/prod. Plus IMHO you will need to reload the code after you remove the dependency... Sounds a bit too much for a rake task.

**Comment by lzap** [comment]
> I am not an expert in our jenkins jobs, but I can't see how can we do it in a rake task - we need to remove code dependency - for dev setups it would be removing the plugin from `gemfile.local.rb`, I am not sure what is the correct step for test/prod. Plus IMHO you will need to reload the code after you remove the dependency... Sounds a bit too much for a rake task.

I'd only test db rollback, that's it.

-- 
Later,
 Lukas #lzap Zapletal

**Comment by ShimShtein** [comment]
I have started a [new gem](https://github.com/ShimShtein/foreman_devel) called `foreman_devel` that should contain all kind of functionality for foreman plugin developers.
My first addition to this gem is exactly this functionality - a generator for creating a migration for foreman plugins. If you think it's a good idea, I will gladly move this gem into foreman's repo and continue to work on it.

**Comment by ezr-ondrej** [comment]
@ShimShtein I'd guess this is no longer needed, right? Plugins do not copy their migrations over to core.

**Comment by ShimShtein** [comment]
There is a section in [how to create a plugin](https://github.com/theforeman/foreman/blob/develop/developer_docs/how_to_create_a_plugin.asciidoc#add-new-migration) file with all relevant information.
This one can be closed.

**Pull request closed without merging**